src/ooni/web_connectivity.cpp

```diff
@@ -0,0 +1,677 @@
+        if (close_txp == true) {
+          txp->close([=]{
+            cb(NoError());
+          });
+        } else {
+          cb(NoError());
+        }
+      } else {
+        if (close_txp == true) {
+          // XXX optimistic closure
+          txp->close([=]{});
+        }
+      }
+    };
+  };
+
+  for (auto socket : sockets) {
+      std::string address = socket.first;
+      int port = socket.second;
+      Settings connect_options;
+      connect_options["host"] = address;
+      connect_options["port"] = port;
+      connect_options["net/timeout"] = 0.2;
+      templates::tcp_connect(connect_options,
+                             handle_connect(address, port),
+                             reactor, logger);
+  }
+}
+
+static void experiment_dns_query(
+        Var<Entry> entry,
+        std::string hostname,
+        std::string nameserver,
+        Callback<Error, std::vector<std::string>> callback,
+        Settings options,
+        Var<Reactor> reactor, Var<Logger> logger) {
+
+  templates::dns_query(entry, "A", "IN", hostname, nameserver,
+        [=](Error err, dns::Message message) {
+          std::vector<std::string> addresses;
+          logger->debug("web_connectivity: experiment_dns_query got response");
+          if (err) {
+            callback(err, addresses);
+            return;
+          }
+          for (auto answer : message.answers) {
+            if (answer.ipv4 != "") {
+              addresses.push_back(answer.ipv4);
+            } else if (answer.hostname != "") {
+              addresses.push_back(answer.ipv4);
+            }
+          }
+          callback(NoError(), addresses);
+        }, options, reactor, logger);
+}
+
+void web_connectivity(std::string input, Settings options,
+                      Callback<Var<Entry>> callback,
+                      Var<Reactor> reactor, Var<Logger> logger) {
+    options["http/max_redirects"] = 20;
+    Var<Entry> entry(new Entry);
+    // This is set from ooni test
+    // (*entry)["client_resolver"] = nullptr;
+    (*entry)["retries"] = nullptr;
+
+    (*entry)["dns_consistency"] = nullptr;
+    (*entry)["body_length_match"] = nullptr;
+    (*entry)["headers_match"] = nullptr;
+    (*entry)["status_code_match"] = nullptr;
+    (*entry)["title_match"] = nullptr;
+
+    (*entry)["accessible"] = nullptr;
+    (*entry)["blocking"] = nullptr;
+
+    (*entry)["control_failure"] = nullptr;
+    (*entry)["http_experiment_failure"] = nullptr;
+    (*entry)["dns_experiment_failure"] = nullptr;
+
+    (*entry)["tcp_connect"] = Entry::array();
+    (*entry)["control"] = Entry({});
+
+    ErrorOr<http::Url> url =
+      mk::http::parse_url_noexcept(input);
+
+    if (!url) {
+      logger->debug("Invalid test url.");
+      callback(entry);
+      return;
+    }
+
+    // XXX check if this hostname is actually a hostname and not an IP
+    std::string hostname = url->address;
+    // XXX we probably want to do like ooni and use the system resolver
```

Comment:
From C++ we don't know the system resolver on mobile devices. Currently Apps must pass us a proper `"dns/nameserver"` pointing to the system resolver found using platform specific methods.

One alternative would be to write a DNS engine that uses `getaddrinfo()` rather than calling libevent's DNS; in such case, the response would come from libc and hence from the system's DNS.